### PR TITLE
reference: Add "Advanced Tagging" section

### PR DESCRIPTION
**Pull request by doanac**
As factories and CI stories get more complex, we are finding the need
to introduce more powerful ways to do handle tagging. This change
adds a new document to help explain what's possible. Its based
on the issue defined here:

 https://github.com/foundriesio/ci-scripts/issues/44

Signed-off-by: Andy Doan <andy@foundries.io>

### CODE REVIEW

**Comment by doanac** [comment]
Docs for e5d513ae9ef5b23d5f3ad53ef00f42a60df045ac are browesable at: https://ci.foundries.io/projects/fio-docs/builds/369/docs/artifacts/html/index.html

**Comment by doanac** [comment]
Rendered page is: https://ci.foundries.io/projects/fio-docs/builds/369/docs/artifacts/html/reference/advanced-tagging.html

File: source/reference/advanced-tagging.rst [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+  TAG[:INHERIT_TAG][,TAG[:INHERIT_TAG]....]
+
+
+Scenario 1: A new platform build that re-uses containers
+--------------------------------------------------------
+
+Consider the case where a factory might have a tag called "postmerge" defined
+for both the lmp.yml and containers.yml builds. A new branch is added to the
+LMP called "postmerge-stable" that's going to be based on older, more stable
```
Comment (EmbeddedAndroid) [inline source/reference/advanced-tagging.rst]: I think these quotes need to be escaped, the rendered version is basically unreadable.

File: source/reference/advanced-tagging.rst [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+The ``OTA_LITE_TAG`` determines what tag a target will get when a container
+or platform build is performed. The format is::
+
+  TAG[:INHERIT_TAG][,TAG[:INHERIT_TAG]....]
+
+
+Scenario 1: A new platform build that re-uses containers
+--------------------------------------------------------
+
+Consider the case where a factory might have a tag called "postmerge" defined
```
Comment (EmbeddedAndroid) [inline source/reference/advanced-tagging.rst]: ditto

File: source/reference/advanced-tagging.rst [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+**Platform Build** - A build created by a change to the LMP/OE. This is the
+base OS image.
+
+**Container Build** - A build created by a change to containers.get.
+
+**Target** - This a TUF entry in a factory's targets.json file. It represents
+what should be thought of as an immutable combination of the Platform build's
+OSTree hash + the output of a Container build.
+
+**Tag** - A Target has a "custom" section where with a list of Tags. The
```
Comment (EmbeddedAndroid) [inline source/reference/advanced-tagging.rst]: ditto

File: source/reference/advanced-tagging.rst [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+Terminology
+-----------
+
+**Platform Build** - A build created by a change to the LMP/OE. This is the
+base OS image.
+
+**Container Build** - A build created by a change to containers.get.
+
+**Target** - This a TUF entry in a factory's targets.json file. It represents
```
Comment (EmbeddedAndroid) [inline source/reference/advanced-tagging.rst]: The apostrophe is being rendered badly as well.

File: source/reference/advanced-tagging.rst [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+Consider the case where a factory might have a tag called "postmerge" defined
+for both the lmp.yml and containers.yml builds. A new branch is added to the
+LMP called "postmerge-stable" that's going to be based on older, more stable
+code. However, this new build will use the containers found in the most
+recent "postmerge" target. This can be expressed in lmp.yml with::
+
+  OTA_LITE_TAG: "postmerge-stable:postmerge"
+
+However, this means that changes to containers.git will now also need to
+produce a new "postmerge-stable" target. So containers.yaml would need to
```
Comment (EmbeddedAndroid) [inline source/reference/advanced-tagging.rst]: here as well

File: source/reference/advanced-tagging.rst [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+
+Scenario 1: A new platform build that re-uses containers
+--------------------------------------------------------
+
+Consider the case where a factory might have a tag called "postmerge" defined
+for both the lmp.yml and containers.yml builds. A new branch is added to the
+LMP called "postmerge-stable" that's going to be based on older, more stable
+code. However, this new build will use the containers found in the most
+recent "postmerge" target. This can be expressed in lmp.yml with::
```
Comment (EmbeddedAndroid) [inline source/reference/advanced-tagging.rst]: ditto

File: source/reference/advanced-tagging.rst [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+      docker-apps: foo:v1, bar:v1
+      tags: postmerge-stable
+    build-2:
+      ostree-hash: GOODBEEF
+      docker-apps: foo:v2, bar:v2
+      tags: postmerge
+
+If a change to the postmerge-stable branch was pushed to the LMP, a new
+target, build-3, would be added. The build logic would then look through
+the targets list to find the most recent "postmerge" target so that
```
Comment (EmbeddedAndroid) [inline source/reference/advanced-tagging.rst]: ditto

File: source/reference/advanced-tagging.rst [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+it can copy those docker-apps. This would result in a new target::
+
+  build-3:
+    ostree-hash: NEWHASH
+    docker-apps: foo:v2, bar:v2
+    tags: postmerge-stable
+
+
+On the other hand, there might also be a new container build for "postmerge".
+In this case the tag specification `postmerge,postmerge-stable` would tell
```
Comment (EmbeddedAndroid) [inline source/reference/advanced-tagging.rst]: Single quotes need to be escaped as well, or use three :)

File: source/reference/advanced-tagging.rst [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    ostree-hash: GOODBEEF
+    docker-apps: foo:v3, bar:v3
+    tags: postmerge
+
+
+Scenario 2: Multiple container builds using the same platform
+-------------------------------------------------------------
+
+This scenario is the reverse of the previous one. A factory might have a
+platform build tagged with "devel-X". However, there are two versions of
```
Comment (EmbeddedAndroid) [inline source/reference/advanced-tagging.rst]: ditto

File: source/reference/advanced-tagging.rst [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+containers being worked on: "devel-X" and "devel-Y". This could be handled
```
Comment (EmbeddedAndroid) [inline source/reference/advanced-tagging.rst]: ditto

File: source/reference/advanced-tagging.rst (hunk 1 above)
Comment (doanac) [reply source/reference/advanced-tagging.rst]: That's just something odd with how jobserv tries to render the page. It looks good when I build it locally.

File: source/reference/advanced-tagging.rst (hunk 8 above)
Comment (doanac) [reply source/reference/advanced-tagging.rst]: that one did need fixing.

**Comment by doanac** [comment]
Docs for 3ac971be47eeda60a5d89207b471e1d520e19202 are browesable at: https://ci.foundries.io/projects/fio-docs/builds/370/docs/artifacts/html/index.html

File: source/reference/advanced-tagging.rst [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+The ``OTA_LITE_TAG`` variable can be used in a few ways to add advanced
+tagging capabilities to a factory.
+
+Terminology
+-----------
+
+**Platform Build** - A build created by a change to the LMP/OE. This is the
+base OS image.
+
+**Container Build** - A build created by a change to containers.get.
```
Comment (ricardosalveti) [inline source/reference/advanced-tagging.rst]: containers.get or git?

File: source/reference/advanced-tagging.rst (hunk 4 above)
Comment (ricardosalveti) [inline source/reference/advanced-tagging.rst]: 'This is a TUF'? Or maybe just 'A TUF entry' to follow the same pattern.

File: source/reference/advanced-tagging.rst (hunk 11 above)
Comment (doanac) [reply source/reference/advanced-tagging.rst]: good catch

File: source/reference/advanced-tagging.rst (hunk 4 above)
Comment (doanac) [reply source/reference/advanced-tagging.rst]: I re-worded it to "This an entry in a factory's TUF targets.json file"

**Comment by doanac** [comment]
Docs for 34b6467bcdcd220ecdb723a40a9320ca94c2c4ed are browesable at: https://ci.foundries.io/projects/fio-docs/builds/371/docs/artifacts/html/index.html

**Comment by doanac** [comment]
Docs for bf36f7cfe829517d77b3b6db5121a70e3da3f014 are browesable at: https://ci.foundries.io/projects/fio-docs/builds/372/docs/artifacts/html/index.html

**Comment by doanac** [comment]
@EmbeddedAndroid @ricardosalveti  - just add a new "Scenario 3" to this PR.

**Comment by doanac** [comment]
Docs for 3c4674b35ca83fd265a5ba2ce0a0b6285854ea33 are browesable at: https://ci.foundries.io/projects/fio-docs/builds/373/docs/artifacts/html/index.html

**Pull request merged**